pkg/container-utils/containerd/containerd.go

```diff
@@ -15,31 +15,281 @@
+	task, err := container.Task(context.TODO(), nil)
+	if err != nil {
+		// According to nerdctl, if there is no task, we can assume the
+		// container was just created but it is not running yet:
+		// https://github.com/containerd/nerdctl/blob/b0a75d880ef9e6d1f1a8752804c9087ee2d02f73/pkg/formatter/formatter.go#L48-L55
+		if !errdefs.IsNotFound(err) {
+			return nil, fmt.Errorf("getting task for container %q: %w", container.ID(), err)
+		}
+
+		t := &containerTask{
+			status: runtimeclient.StateCreated,
+			pid:    0,
+		}
+		log.Debugf("%sNo task for %q. Assuming it is in %q status",
+			logPrefix, container.ID(), t.status)
+
+		return t, nil
+	}
+
+	containerdStatus, err := task.Status(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("getting status of task for container %q: %w", container.ID(), err)
+	}
+
+	return &containerTask{
+		status: processStatusStateToRuntimeClientState(containerdStatus.Status),
+		pid:    task.Pid(),
+	}, nil
+}
+
+// Constructs a ContainerData from a containerTask and containerd.Container
+// The extra containerd.Container parameter saves an additional call to the API
+func taskAndContainerToContainerData(task *containerTask, container containerd.Container) (*runtimeclient.ContainerData, error) {
+	labels, err := container.Labels(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("listing labels of container %q: %w", container.ID(), err)
+	}
+
+	containerData := &runtimeclient.ContainerData{
+		ID:      container.ID(),
+		Name:    getContainerName(container),
+		State:   task.status,
+		Runtime: runtimeclient.ContainerdName,
+	}
+	runtimeclient.EnrichWithK8sMetadata(containerData, labels)
+	return containerData, nil
+}
+
+// Checks if the K8s Label for the Containerkind equals to sandbox
+func isSandboxContainer(container containerd.Container) bool {
+	labels, err := container.Labels(context.TODO())
+	if err != nil {
+		return false
+	}
+
+	if kind, ok := labels[LabelK8sContainerdKind]; ok {
+		return kind == LabelK8sContainerdKindSandbox
+	}
+
+	return false
+}
+
+// Convert the state from container status to state of runtime client.
+func processStatusStateToRuntimeClientState(status containerd.ProcessStatus) string {
+	switch status {
+	case containerd.Created:
+		return runtimeclient.StateCreated
+	case containerd.Running:
+		return runtimeclient.StateRunning
+	case containerd.Stopped:
+		return runtimeclient.StateExited
+	default:
+		return runtimeclient.StateUnknown
+	}
+}
+
+// getContainerName returns the name of the container. If the container is
+// managed by Kubernetes, it returns the name of the container as defined in
+// Kubernetes. Otherwise, it returns the container ID.
+func getContainerName(container containerd.Container) string {
+	labels, err := container.Labels(context.TODO())
+	if err != nil {
+		return container.ID()
+	}
+
+	if k8sName, ok := labels[LabelK8sContainerName]; ok {
+		return k8sName
+	}
+
```

Comment:
nit (for another PR): We could also look for `nerdctl/name` to support containers created with nerdctl.